README
-------------------------------------------------------------------------------
README
-------------------------------------------------------------------------------

DSLogic-fw is an Open-source firmware for DSLogic
DSLogic project: www.dreamsourcelab.com

Status
------

DSLogic-fw is in a usable state and has had official
tarball releases.


Copyright and license
---------------------

DSLogic-firmware is licensed under the terms of the GNU General Public
License (GPL), version 2 or later.

It uses additional library EZUSB.lib, provided by CPRESS (http://www.cypress.com).


Website
-------

 http://www.dreamsourcelab.com



NEWS
1.0 (2014-07-21)
------------------

 * Initial release.



Source/fw.c
//-----------------------------------------------------------------------------
//   File:      fw.c
//   Contents:  Firmware frameworks task dispatcher and device request parser
//              source.
//
// indent 3.  NO TABS!
//
// $Revision: 17 $
// $Date: 11/15/01 5:45p $
//
//   Copyright (c) 2003 Cypress Semiconductor, Inc. All rights reserved
//-----------------------------------------------------------------------------
#include "include/fx2.h"
#include "include/fx2regs.h"

//-----------------------------------------------------------------------------
// Random Macros
//-----------------------------------------------------------------------------
#define   min(a,b) (((a)<(b))?(a):(b))
#define   max(a,b) (((a)>(b))?(a):(b))

#include "include/syncdly.h"             // Define _IFREQ and _CFREQ above this #include

//-----------------------------------------------------------------------------
// Global Variables
//-----------------------------------------------------------------------------
volatile BOOL   GotSUD;
BOOL   Rwuen;
BOOL   Selfpwr;

WORD   pDeviceDscr;   // Pointer to Device Descriptor; Descriptors may be moved
WORD   pDeviceQualDscr;
WORD   pHighSpeedConfigDscr;
WORD   pFullSpeedConfigDscr;   
WORD   pConfigDscr;
WORD   pOtherConfigDscr;   
WORD   pStringDscr;   

//-----------------------------------------------------------------------------
// Prototypes
//-----------------------------------------------------------------------------
void SetupCommand(void);
void DSLogic_Init(void);
void DSLogic_Poll(void);

BOOL DR_GetDescriptor(void);
BOOL DR_SetConfiguration(void);
BOOL DR_GetConfiguration(void);
BOOL DR_SetInterface(void);
BOOL DR_GetInterface(void);
BOOL DR_GetStatus(void);
BOOL DR_ClearFeature(void);
BOOL DR_SetFeature(void);
BOOL DR_VendorCmnd(void);

// this table is used by the epcs macro 
const char code  EPCS_Offset_Lookup_Table[] =
{
   0,    // EP1OUT
   1,    // EP1IN
   2,    // EP2OUT
   2,    // EP2IN
   3,    // EP4OUT
   3,    // EP4IN
   4,    // EP6OUT
   4,    // EP6IN
   5,    // EP8OUT
   5,    // EP8IN
};

// macro for generating the address of an endpoint's control and status register (EPnCS)
#define epcs(EP) (EPCS_Offset_Lookup_Table[(EP & 0x7E) | (EP > 128)] + 0xE6A1)

//-----------------------------------------------------------------------------
// Code
//-----------------------------------------------------------------------------

// Task dispatcher
void main(void)
{
   DWORD   i;
   WORD   offset;
   DWORD   DevDescrLen;
   DWORD   j=0;
   WORD   IntDescrAddr;
   WORD   ExtDescrAddr;

   // Initialize Global States
   Rwuen = FALSE;           // Disable remote wakeup
   Selfpwr = FALSE;         // Disable self powered
   GotSUD = FALSE;          // Clear "Got setup data" flag

   // Initialize user device
   DSLogic_Init();

   // Relocate the descriptor table. 
   pDeviceDscr = (WORD)&DeviceDscr;
   pDeviceQualDscr = (WORD)&DeviceQualDscr;
   pHighSpeedConfigDscr = (WORD)&HighSpeedConfigDscr;
   pFullSpeedConfigDscr = (WORD)&FullSpeedConfigDscr;
   pStringDscr = (WORD)&StringDscr;

   if (EZUSB_HIGHSPEED())
   {
      pConfigDscr = pHighSpeedConfigDscr;
      pOtherConfigDscr = pFullSpeedConfigDscr;
   }
   else
   {
      pConfigDscr = pFullSpeedConfigDscr;
      pOtherConfigDscr = pHighSpeedConfigDscr;
   }

   if ((WORD)&DeviceDscr & 0xe000)
   {
      IntDescrAddr = INTERNAL_DSCR_ADDR;
      ExtDescrAddr = (WORD)&DeviceDscr;
      DevDescrLen = (WORD)&UserDscr - (WORD)&DeviceDscr + 2;
      for (i = 0; i < DevDescrLen; i++)
         *((BYTE xdata *)IntDescrAddr+i) = 0xCD;
      for (i = 0; i < DevDescrLen; i++)
         *((BYTE xdata *)IntDescrAddr+i) = *((BYTE xdata *)ExtDescrAddr+i);
      pDeviceDscr = IntDescrAddr;
      offset = (WORD)&DeviceDscr - INTERNAL_DSCR_ADDR;
      pDeviceQualDscr -= offset;
      pConfigDscr -= offset;
      pOtherConfigDscr -= offset;
      pHighSpeedConfigDscr -= offset;
      pFullSpeedConfigDscr -= offset;
      pStringDscr -= offset;
   }

   // Enable USB interrupt (INT2)
   EZUSB_IRQ_ENABLE();            
   // Enable INT 2 & 4 autovectoring
   INTSETUP |= (bmAV2EN | bmAV4EN);     

   // Enalbe Setup Data Available/Setup Token/SOF/USB Reset/High Speed/ interrupts
   USBIE |= bmSUDAV | bmSUTOK | bmSOF | bmURES | bmHSGRANT;
   // Enable 8051 interrupts
   EA = 1;

   // Unconditionally Renumerate
   EZUSB_Discon(TRUE);
   USBCS &=~bmDISCON;
   
   // Set stretch to 0 (after renumeration)
   CKCON = (CKCON&(~bmSTRETCH)) | FW_STRETCH_VALUE; 

   // Task Dispatcher
   while(TRUE)
   {
      if(GotSUD)
      {
         SetupCommand();	// Implement setup command
         GotSUD = FALSE;	// Clear SUDAV flag
      }

      DSLogic_Poll();
   }
}

// Device request parser
void SetupCommand(void)
{
   void   *dscr_ptr;

   switch(SETUPDAT[1])
   {
      case SC_GET_DESCRIPTOR:                  // *** Get Descriptor
         if(DR_GetDescriptor())
            switch(SETUPDAT[3])         
            {
               case GD_DEVICE:            // Device
                  SUDPTRH = MSB(pDeviceDscr);
                  SUDPTRL = LSB(pDeviceDscr);
                  break;
               case GD_DEVICE_QUALIFIER:            // Device Qualifier
                  SUDPTRH = MSB(pDeviceQualDscr);
                  SUDPTRL = LSB(pDeviceQualDscr);
                  break;
               case GD_CONFIGURATION:         // Configuration
                  SUDPTRH = MSB(pConfigDscr);
                  SUDPTRL = LSB(pConfigDscr);
                  break;
               case GD_OTHER_SPEED_CONFIGURATION:  // Other Speed Configuration
                  SUDPTRH = MSB(pOtherConfigDscr);
                  SUDPTRL = LSB(pOtherConfigDscr);
                  break;
               case GD_STRING:            // String
                  if(dscr_ptr = (void *)EZUSB_GetStringDscr(SETUPDAT[2]))
                  {
                     SUDPTRH = MSB(dscr_ptr);
                     SUDPTRL = LSB(dscr_ptr);
                  }
                  else 
                     EZUSB_STALL_EP0();   // Stall End Point 0
                  break;
               default:            // Invalid request
                  EZUSB_STALL_EP0();      // Stall End Point 0
            }
         break;
      case SC_GET_INTERFACE:                  // *** Get Interface
         DR_GetInterface();
         break;
      case SC_SET_INTERFACE:                  // *** Set Interface
         DR_SetInterface();
         break;
      case SC_SET_CONFIGURATION:               // *** Set Configuration
         DR_SetConfiguration();
         break;
      case SC_GET_CONFIGURATION:               // *** Get Configuration
         DR_GetConfiguration();
         break;
      case SC_GET_STATUS:                  // *** Get Status
         if(DR_GetStatus())
            switch(SETUPDAT[0])
            {
               case GS_DEVICE:            // Device
                  EP0BUF[0] = ((BYTE)Rwuen << 1) | (BYTE)Selfpwr;
                  EP0BUF[1] = 0;
                  EP0BCH = 0;
                  EP0BCL = 2;
                  break;
               case GS_INTERFACE:         // Interface
                  EP0BUF[0] = 0;
                  EP0BUF[1] = 0;
                  EP0BCH = 0;
                  EP0BCL = 2;
                  break;
               case GS_ENDPOINT:         // End Point
                  EP0BUF[0] = *(BYTE xdata *) epcs(SETUPDAT[4]) & bmEPSTALL;
                  EP0BUF[1] = 0;
                  EP0BCH = 0;
                  EP0BCL = 2;
                  break;
               default:            // Invalid Command
                  EZUSB_STALL_EP0();      // Stall End Point 0
            }
         break;
      case SC_CLEAR_FEATURE:                  // *** Clear Feature
         if(DR_ClearFeature())
            switch(SETUPDAT[0])
            {
               case FT_DEVICE:            // Device
                  if(SETUPDAT[2] == 1)
                     Rwuen = FALSE;       // Disable Remote Wakeup
                  else
                     EZUSB_STALL_EP0();   // Stall End Point 0
                  break;
               case FT_ENDPOINT:         // End Point
                  if(SETUPDAT[2] == 0)
                  {
                     *(BYTE xdata *) epcs(SETUPDAT[4]) &= ~bmEPSTALL;
                     EZUSB_RESET_DATA_TOGGLE( SETUPDAT[4] );
                  }
                  else
                     EZUSB_STALL_EP0();   // Stall End Point 0
                  break;
            }
         break;
      case SC_SET_FEATURE:                  // *** Set Feature
         if(DR_SetFeature())
            switch(SETUPDAT[0])
            {
               case FT_DEVICE:            // Device
                  if(SETUPDAT[2] == 1)
                     Rwuen = TRUE;      // Enable Remote Wakeup
                  else if(SETUPDAT[2] == 2)
                     // Set Feature Test Mode.  The core handles this request.  However, it is
                     // necessary for the firmware to complete the handshake phase of the
                     // control transfer before the chip will enter test mode.  It is also
                     // necessary for FX2 to be physically disconnected (D+ and D-)
                     // from the host before it will enter test mode.
                     break;
                  else
                     EZUSB_STALL_EP0();   // Stall End Point 0
                  break;
               case FT_ENDPOINT:         // End Point
                  *(BYTE xdata *) epcs(SETUPDAT[4]) |= bmEPSTALL;
                  break;
            }
         break;
      default:                     // *** Verdor Command
         if(DR_VendorCmnd())
            EZUSB_STALL_EP0();            // Stall End Point 0
   }

   // Acknowledge handshake phase of device request
   EP0CS |= bmHSNAK;
}

// Wake-up interrupt handler
void resume_isr(void) interrupt WKUP_VECT
{
   EZUSB_CLEAR_RSMIRQ();
}




Source/include/DSLogic.h
/*
 * This file is part of the DSLogic-firmware project.
 *
 * Copyright (C) 2014 DreamSourceLab <support@dreamsourcelab.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
 */

#ifndef DSLOGICFW_INCLUDE_DSLOGIC_H
#define DSLOGICFW_INCLUDE_DSLOGIC_H

/*
 * Major and minor DSLogic-firmware version numbers.
 * These can be queried by the host via CMD_GET_FW_VERSION.
 */
#define DSLOGICFW_VER_MAJOR	1
#define DSLOGICFW_VER_MINOR	0

/* Protocol commands */
#define CMD_GET_FW_VERSION		0xb0
#define CMD_GET_REVID_VERSION		0xb1
#define CMD_START			0xb2
#define CMD_CONFIG			0xb3
#define CMD_SETTING			0xb4
#define CMD_CONTROL			0xb5

#define CMD_START_FLAGS_WIDE_POS	5
#define CMD_START_FLAGS_CLK_SRC_POS	6
#define CMD_START_FLAGS_STOP_POS	7

#define CMD_START_FLAGS_SAMPLE_8BIT	(0 << CMD_START_FLAGS_WIDE_POS)
#define CMD_START_FLAGS_SAMPLE_16BIT	(1 << CMD_START_FLAGS_WIDE_POS)

#define CMD_START_FLAGS_CLK_30MHZ	(0 << CMD_START_FLAGS_CLK_SRC_POS)
#define CMD_START_FLAGS_CLK_48MHZ	(1 << CMD_START_FLAGS_CLK_SRC_POS)

#define CMD_START_FLAGS_STOP		(1 << CMD_START_FLAGS_STOP_POS)

struct cmd_start {
	BYTE flags;
	BYTE param0;
	BYTE param1;
};

struct cmd_setting_count {
	BYTE byte0;
	BYTE byte1;
	BYTE byte2;
};

struct cmd_cfg_count {
	BYTE byte0;
	BYTE byte1;
	BYTE byte2;
};

struct cmd_control {
    	BYTE byte0;
    	BYTE byte1;
    	BYTE byte2;
    	BYTE byte3;
};

#endif


Source/DSLogic.c
#pragma NOIV               // Do not generate interrupt vectors
/*
 * This file is part of the DSLogic-firmware project.
 *
 * Copyright (C) 2014 DreamSourceLab <support@dreamsourcelab.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
 */

#include "include/fx2.h"
#include "include/fx2regs.h"
#include "include/syncdly.h"    // SYNCDELAY macro, see Section 15.14 of FX2 Tech.
                                // Ref. Manual for usage details.
#include "include/DSLogic.h"

#define GPIFTRIGWR 0
#define GPIFTRIGRD 4

#define GPIF_EP2 0
#define GPIF_EP4 1
#define GPIF_EP6 2
#define GPIF_EP8 3

extern BOOL GotSUD;             // Received setup data flag
extern BOOL Rwuen;
extern BOOL Selfpwr;

BYTE Configuration;                 // Current configuration
BYTE AlternateSetting;              // Alternate settings
BOOL in_enable = FALSE;             // flag to enable IN transfers
BOOL enum_high_speed = FALSE;       // flag to let firmware know FX2 enumerated at high speed
BOOL cfg_enable = FALSE;
BOOL cfg_init = FALSE;
BOOL set_enable = FALSE;
BOOL set_dso_ctrl = FALSE;
BYTE command = 0;
BYTE setting_count_b0;
BYTE setting_count_b1;
BYTE setting_count_b2;
BYTE xdata dsoConfig[4] = {0x00, 0xc8, 0x61, 0x55};

//-----------------------------------------------------------------------------
// Task Dispatcher hooks
//   The following hooks are called by the task dispatcher.
//-----------------------------------------------------------------------------

void setup_gpif_waveforms(void);
void init_capture_intf(void);
void init_config_intf(void);
void start_capture(void);
void stop_capture(void);
void poll_intf(void);

void DSLogic_Init(void)
{
  // CLKSPD[1:0]=10, for 48MHz operation
  CPUCS = 0x12;                 
  SYNCDELAY;
  
  // Setup Endpoints
  SYNCDELAY;  
  EP2CFG = 0xA0;     // EP2OUT, bulk, size 512, 2x buffered
  SYNCDELAY;           
  EP6CFG = 0xE0;     // EP6IN, bulk, size 512, 2x buffered
  SYNCDELAY;
  
  // Reset FIFOs
  FIFORESET = 0x80;  // set NAKALL bit to NAK all transfers from host
  SYNCDELAY;
  FIFORESET = 0x02;  // reset EP2 FIFO
  SYNCDELAY;
  FIFORESET = 0x06;  // reset EP6 FIFO
  SYNCDELAY;
  FIFORESET = 0x00;  // clear NAKALL bit to resume normal operation
  SYNCDELAY;

  // EP2 Configuration
  EP2FIFOCFG = 0x00; 	// allow core to see zero to one transition of auto out bit
  SYNCDELAY;
  EP2FIFOCFG = 0x10; 	// auto out mode, disable PKTEND zero length send, word ops
  SYNCDELAY;
  EP2GPIFFLGSEL = 0x01; // For EP2OUT, GPIF uses Empty flag
  SYNCDELAY;

  // EP6 Configuration
  EP6FIFOCFG = 0x08; 	// auto in mode, disable PKTEND zero length send, word ops
  SYNCDELAY;
  EP6AUTOINLENH = 0x02;	// Auto-commit 512 (0x200) byte packets (due to AUTOIN = 1)
  SYNCDELAY;
  EP6AUTOINLENL = 0x00;
  SYNCDELAY;
  EP6GPIFFLGSEL = 0x02;	// For EP6IN, Set the GPIF flag to 'full'
  SYNCDELAY;
  
  // PA1~PA0: LED; PA2: Sample Enable; PA3: Sample Clear
  OEA = 0x0f;
  IOA = 0x00;
  
  // GPIF Wavedata Init
  setup_gpif_waveforms( );

  // I2C Interface Init
  EZUSB_InitI2C();
}

void DSLogic_Poll(void)
{
  if (cfg_init && EP0BCL == sizeof(struct cmd_cfg_count))
  {
    cfg_init = FALSE;
    init_config_intf();   
    FIFORESET = 0x80;  // set NAKALL bit to NAK all transfers from host
    SYNCDELAY;
    FIFORESET = 0x02;  // reset EP2 FIFO
    SYNCDELAY;
    FIFORESET = 0x06;  // reset EP6 FIFO
    SYNCDELAY;
    FIFORESET = 0x00;  // clear NAKALL bit to resume normal operation
    SYNCDELAY;	 

    EP2FIFOCFG = 0x00; // allow core to see zero to one transition of auto out bit
    SYNCDELAY;
    EP2FIFOCFG = 0x10; // auto out mode, disable PKTEND zero length send, word ops
    SYNCDELAY;
    EP6FIFOCFG = 0x08; // auto in mode, disable PKTEND zero length send, word ops
    SYNCDELAY;   	 

    GPIFIDLECTL &= 0xFB;	//PROG_B signal low
    EZUSB_Delay1ms();		//PROG_B signal kept asserted for 1 ms
    GPIFIDLECTL |= 0x04;	//PROG_B signal high
    SYNCDELAY;

    // setup transaction count
    GPIFTCB0 = ((const struct cmd_cfg_count *)EP0BUF)->byte0;   
    SYNCDELAY;
    GPIFTCB1 = ((const struct cmd_cfg_count *)EP0BUF)->byte1;            		
    SYNCDELAY;
    GPIFTCB2 = ((const struct cmd_cfg_count *)EP0BUF)->byte2;
    SYNCDELAY;

    cfg_enable = TRUE;			
  }  
  
  if (cfg_enable && (GPIFTRIG & 0x80)) 		// if GPIF interface IDLE
  {        		
    if ( (EP24FIFOFLGS & 0x01) && (GPIFREADYSTAT & 0x01)) {
      // if there's a packet in the peripheral domain for EP2
      // and FPGA is ready to receive the configuration bitstream
      IFCONFIG = 0xA6;
      // 7	IFCLKSRC=1   , FIFOs executes on internal clk source
      // 6	xMHz=0       , 30MHz internal clk rate
      // 5	IFCLKOE=1    , Drive IFCLK pin signal at 30MHz
      // 4	IFCLKPOL=0   , Don't invert IFCLK pin signal from internal clk
      // 3	ASYNC=0      , master samples asynchronous
      // 2	GSTATE=1     , Drive GPIF states out on PORTE[2:0], debug WF
      // 1:0	IFCFG=10, FX2 in GPIF master mode
      SYNCDELAY;

      //delay(1);				//avoid CSI_B deasserted during sync words
      GPIFTRIG = GPIFTRIGWR | GPIF_EP2;  	// launch GPIF FIFO WRITE Transaction from EP2 FIFO
      SYNCDELAY;

      while( !( GPIFTRIG & 0x80 ) );      	// poll GPIFTRIG.7 GPIF Done bit
      SYNCDELAY;
      cfg_enable= FALSE;                 	//end of configuration

      /* Put the FX2 into GPIF master mode and setup the GPIF. */
      //init_capture_intf();

      if (GPIFREADYSTAT & 0x02) {	// FPGA Configure Done
        IOA |= 0x01;
        IOA &= 0xf5;
        EZUSB_Delay1ms();
        IOA |= 0x08;
      } else {
        IOA &= 0xfc;
      }
    }	
  }

  switch(command)
  {
    case CMD_START:
	{
      if ((EP0CS & bmEPBUSY) != 0)
        break;
      if (EP0BCL == sizeof(struct cmd_start))
	  {
        if ((*(BYTE *)EP0BUF) & CMD_START_FLAGS_STOP)
          stop_capture();
        else
          start_capture();
      }
      command = 0;
      break;	
	}

	case CMD_SETTING:
    {
	  if ((EP0CS & bmEPBUSY) != 0)
        break;
 	  if (EP0BCL == sizeof(struct cmd_setting_count))
	  {
	    GPIFABORT = 0xff;
		SYNCDELAY;
		EP2FIFOCFG = 0x11; // auto out mode, disable PKTEND zero length send, word operation
		SYNCDELAY;
        setting_count_b0 = ((const struct cmd_setting_count *)EP0BUF)->byte0;
        setting_count_b1 = ((const struct cmd_setting_count *)EP0BUF)->byte1;
        setting_count_b2 = ((const struct cmd_setting_count *)EP0BUF)->byte2;
        set_enable = TRUE;
	  }
	  command = 0;
	  break;
    }

    case CMD_CONTROL:
    {
      if ((EP0CS & bmEPBUSY) != 0)
        break;
      if (EP0BCL == sizeof(struct cmd_control))
	  {
        dsoConfig[0] = ((const struct cmd_control *)EP0BUF)->byte0;
        dsoConfig[1] = ((const struct cmd_control *)EP0BUF)->byte1;
        dsoConfig[2] = ((const struct cmd_control *)EP0BUF)->byte2;
        dsoConfig[3] = ((const struct cmd_control *)EP0BUF)->byte3;
        set_dso_ctrl = TRUE;
      }
	  command = 0;
	  break;
    }

	default:
	  command = 0;
	  break;
  }

  if (set_enable && (GPIFTRIG & 0x80)) {	// if GPIF interface IDLE
    if (!(EP24FIFOFLGS & 0x02)) {
      SYNCDELAY;
      GPIFTCB2 = setting_count_b2;   
      SYNCDELAY;
      GPIFTCB1 = setting_count_b1;			// fpga setting count
      SYNCDELAY;
      GPIFTCB0 = setting_count_b0;
      SYNCDELAY;

      GPIFTRIG = GPIFTRIGWR | GPIF_EP2;  	// launch GPIF FIFO WRITE Transaction from EP2 FIFO
      SYNCDELAY;

      while( !( GPIFTRIG & 0x80 ) );      	// poll GPIFTRIG.7 GPIF Done bit
      SYNCDELAY;
      set_enable= FALSE;                 	//end of configuration

      /* Put the FX2 into GPIF master mode and setup the GPIF. */
      init_capture_intf();
    }	
  }

  if (set_dso_ctrl) {
    EZUSB_WriteI2C(0x51, 4, dsoConfig);
    set_dso_ctrl = FALSE;
  }

  poll_intf();  
}

//-----------------------------------------------------------------------------
// Device Request hooks
//   The following hooks are called by the end point 0 device request parser.
//-----------------------------------------------------------------------------

BOOL DR_GetDescriptor(void)
{
   return(TRUE);
}

BOOL DR_SetConfiguration(void)   // Called when a Set Configuration command is received
{
  if( EZUSB_HIGHSPEED( ) )
  { // FX2 enumerated at high speed
    SYNCDELAY;                  // 
    EP6AUTOINLENH = 0x02;       // set AUTOIN commit length to 512 bytes
    SYNCDELAY;  
    EP6AUTOINLENL = 0x00;
    SYNCDELAY;                 // 
                      
    enum_high_speed = TRUE;
  }
  else
  { // FX2 enumerated at full speed
    SYNCDELAY;                   
    EP6AUTOINLENH = 0x00; 
    SYNCDELAY;                 
    EP6AUTOINLENL = 0x40;
    SYNCDELAY;                  
    enum_high_speed = FALSE;
  }

  Configuration = SETUPDAT[2];
  return(TRUE);            // Handled by user code
}

BOOL DR_GetConfiguration(void)   // Called when a Get Configuration command is received
{
   EP0BUF[0] = Configuration;
   EP0BCH = 0;
   EP0BCL = 1;
   return(TRUE);            // Handled by user code
}

BOOL DR_SetInterface(void)       // Called when a Set Interface command is received
{
   AlternateSetting = SETUPDAT[2];
   return(TRUE);            // Handled by user code
}

BOOL DR_GetInterface(void)       // Called when a Set Interface command is received
{
   EP0BUF[0] = AlternateSetting;
   EP0BCH = 0;
   EP0BCL = 1;
   return(TRUE);            // Handled by user code
}

BOOL DR_GetStatus(void)
{
   return(TRUE);
}

BOOL DR_ClearFeature(void)
{
   return(TRUE);
}

BOOL DR_SetFeature(void)
{
   return(TRUE);
}

BOOL DR_VendorCmnd(void)
{
  switch (SETUPDAT[1])
  {
    case CMD_GET_FW_VERSION:
	{
	  EP0BUF[0] = DSLOGICFW_VER_MAJOR;
	  EP0BUF[1] = DSLOGICFW_VER_MINOR;
	  EP0BCH = 0;
	  EP0BCL = 2;
	  break;
	}
    case CMD_GET_REVID_VERSION:
	{
	  EP0BUF[0] = REVID;
	  EP0BCH = 0;
	  EP0BCL = 1;
	  break;
	}
    case CMD_START:
	case CMD_SETTING:
	case CMD_CONTROL:
    {
      command = SETUPDAT[1];
      EP0BCL = 0;
	  break;
    }

    case CMD_CONFIG:
    {
	  EP0BCL = 0;
	  IOA |= 0x03;
	  cfg_init = TRUE;
	  break;
    }

    default:
      return(TRUE);
  }

  return(FALSE);
}

//-----------------------------------------------------------------------------
// USB Interrupt Handlers
//   The following functions are called by the USB interrupt jump table.
//-----------------------------------------------------------------------------

// Setup Data Available Interrupt Handler
void ISR_Sudav(void) interrupt 0
{
   GotSUD = TRUE;            // Set flag
   EZUSB_IRQ_CLEAR();
   USBIRQ = bmSUDAV;         // Clear SUDAV IRQ
}

// Setup Token Interrupt Handler
void ISR_Sutok(void) interrupt 0
{
   EZUSB_IRQ_CLEAR();
   USBIRQ = bmSUTOK;         // Clear SUTOK IRQ
}

void ISR_Sof(void) interrupt 0
{
   EZUSB_IRQ_CLEAR();
   USBIRQ = bmSOF;            // Clear SOF IRQ
}

void ISR_Ures(void) interrupt 0
{
   // whenever we get a USB reset, we should revert to full speed mode
   pConfigDscr = pFullSpeedConfigDscr;
   ((CONFIGDSCR xdata *) pConfigDscr)->type = CONFIG_DSCR;
   pOtherConfigDscr = pHighSpeedConfigDscr;
   ((CONFIGDSCR xdata *) pOtherConfigDscr)->type = OTHERSPEED_DSCR;

   EZUSB_IRQ_CLEAR();
   USBIRQ = bmURES;         // Clear URES IRQ
}

void ISR_Susp(void) interrupt 0
{
}

void ISR_Highspeed(void) interrupt 0
{
   if (EZUSB_HIGHSPEED())
   {
      pConfigDscr = pHighSpeedConfigDscr;
      ((CONFIGDSCR xdata *) pConfigDscr)->type = CONFIG_DSCR;
      pOtherConfigDscr = pFullSpeedConfigDscr;
      ((CONFIGDSCR xdata *) pOtherConfigDscr)->type = OTHERSPEED_DSCR;
   }

   EZUSB_IRQ_CLEAR();
   USBIRQ = bmHSGRANT;
}
void ISR_Ep0ack(void) interrupt 0
{
}
void ISR_Stub(void) interrupt 0
{
}
void ISR_Ep0in(void) interrupt 0
{
}
void ISR_Ep0out(void) interrupt 0
{
}
void ISR_Ep1in(void) interrupt 0
{
}
void ISR_Ep1out(void) interrupt 0
{
}
void ISR_Ep2inout(void) interrupt 0
{
}
void ISR_Ep4inout(void) interrupt 0
{
}
void ISR_Ep6inout(void) interrupt 0
{
}
void ISR_Ep8inout(void) interrupt 0
{
}
void ISR_Ibn(void) interrupt 0
{
}
void ISR_Ep0pingnak(void) interrupt 0
{
}
void ISR_Ep1pingnak(void) interrupt 0
{
}
void ISR_Ep2pingnak(void) interrupt 0
{
}
void ISR_Ep4pingnak(void) interrupt 0
{
}
void ISR_Ep6pingnak(void) interrupt 0
{
}
void ISR_Ep8pingnak(void) interrupt 0
{
}
void ISR_Errorlimit(void) interrupt 0
{
}
void ISR_Ep2piderror(void) interrupt 0
{
}
void ISR_Ep4piderror(void) interrupt 0
{
}
void ISR_Ep6piderror(void) interrupt 0
{
}
void ISR_Ep8piderror(void) interrupt 0
{
}
void ISR_Ep2pflag(void) interrupt 0
{
}
void ISR_Ep4pflag(void) interrupt 0
{
}
void ISR_Ep6pflag(void) interrupt 0
{
}
void ISR_Ep8pflag(void) interrupt 0
{
}
void ISR_Ep2eflag(void) interrupt 0
{
}
void ISR_Ep4eflag(void) interrupt 0
{
}
void ISR_Ep6eflag(void) interrupt 0
{
}
void ISR_Ep8eflag(void) interrupt 0
{
}
void ISR_Ep2fflag(void) interrupt 0
{
}
void ISR_Ep4fflag(void) interrupt 0
{
}
void ISR_Ep6fflag(void) interrupt 0
{
}
void ISR_Ep8fflag(void) interrupt 0
{
}
void ISR_GpifComplete(void) interrupt 0
{
}
void ISR_GpifWaveform(void) interrupt 0
{
}
